### PR TITLE
`snlls`: Fix violation of boundaries due to float-point round-off errors 

### LINKED ISSUE #177: Bounds violation error in `fitmultimodel`
With the following MWE I get a hard to reproduce error: 
```python
import numpy as np 
import matplotlib.pyplot as plt 
import deerlab as dl

data = np.loadtxt('ground_truth_23.dat')
t = data[:,0]/1000 # ns->us
V = data[:,1]

t = dl.correctzerotime(V,t)
rmin = 1.2512846400919055
rmax = 9.59687398219514
r = np.linspace(rmin,rmax,200)

Kmodel = lambda p: dl.dipolarkernel(t,r,mod=p[0],bg=dl.bg_hom3d(t,p[1],p[0]))
fit = dl.fitmultimodel(V,Kmodel,r,dl.dd_gauss,4,lbK=[0,0],ubK=[1,500])
fit.plot()
```
Here is the dataset required for the example to run:
[ground_truth_23.zip](https://github.com/JeschkeLab/DeerLab/files/6492412/ground_truth_23.zip)

and here is the error callback: 
```
ValueError                                Traceback (most recent call last)
d:\lufa\projects\denoising\test_set_100\test_set_100\analysis_MG_denoising_data.py in 
     106 
     107 Kmodel = lambda p: dl.dipolarkernel(t,r,mod=p[0],bg=dl.bg_hom3d(t,p[1],p[0]))
---> 108 fit = dl.fitmultimodel(V,Kmodel,r,dl.dd_gauss,4,lbK=[0,0],ubK=[1,500])
     109 fit.plot()

d:\lufa\projects\deerlab\deerlab\deerlab\fitmultimodel.py in fitmultimodel(V, Kmodel, r, model, maxModels, method, lb, ub, lbK, ubK, strategy, weights, renormalize, uq, tol, maxiter)
    427         # Separable non-linear least-squares (SNLLS) fit
    428         upscale = 1e2
--> 429         fit = dl.snlls(V*upscale,Knonlin,par0,nlin_lb,nlin_ub,lin_lb,lin_ub, 
    430                         weights=weights, reg=False, nonlin_tol=tol, nonlin_maxiter=maxiter)
    431         pnonlin = fit.nonlin

d:\lufa\projects\deerlab\deerlab\deerlab\snlls.py in snlls(y, Amodel, par0, lb, ub, lbl, ubl, nnlsSolver, reg, weights, regtype, regparam, multistart, regorder, alphareopt, extrapenalty, nonlin_tol, nonlin_maxiter, lin_tol, lin_maxiter, huberparam, uq)
    459         nonlin_subset = np.arange(0,Nnonlin)
    460         lin_subset = np.arange(Nnonlin,Nnonlin+Nlin)
--> 461         paramuq_nonlin = uq_subset(paramuq,nonlin_subset,lb,ub)
    462         paramuq_lin = uq_subset(paramuq,lin_subset,lbl,ubl)
    463 

d:\lufa\projects\deerlab\deerlab\deerlab\snlls.py in uq_subset(uq_full, subset, subset_lb, subset_ub)
    511 
    512     subset_model = lambda x: x[subset]
--> 513     uq_subset = uq_full.propagate(subset_model,lbm=subset_lb, ubm=subset_ub)
    514 
    515     return uq_subset

d:\lufa\projects\deerlab\deerlab\deerlab\classes.py in propagate(self, model, lbm, ubm)
    386 
    387         # Get jacobian of model to be propagated with respect to parameters
--> 388         J = Jacobian(model,parfit,self.__lb,self.__ub)
    389 
    390         # Clip at boundaries

d:\lufa\projects\deerlab\deerlab\deerlab\utils\utils.py in Jacobian(fcn, x0, lb, ub)
    409 
    410     """
--> 411     J = opt._numdiff.approx_derivative(fcn,x0,method='2-point',bounds=(lb,ub))
    412     J = np.atleast_2d(J)
    413     return J

~\AppData\Local\Programs\Python\Python38\lib\site-packages\scipy\optimize\_numdiff.py in approx_derivative(fun, x0, method, rel_step, abs_step, f0, bounds, sparsity, as_linear_operator, args, kwargs)
    449 
    450     if np.any((x0 < lb) | (x0 > ub)):
--> 451         raise ValueError("`x0` violates bound constraints.")
    452 
    453     if as_linear_operator:

ValueError: `x0` violates bound constraints.
```

### PR DESCRIPTION
**Pull request by luisfabib**
Closes #177

### CODE REVIEW

File: deerlab/utils/utils.py [hunk 1]
```diff
@@ -408,7 +408,15 @@ def Jacobian(fcn, x0, lb, ub):
     This is a wrapper around the ``scipy.optimize._numdiff.approx_derivative`` function.
 
     """
-    J = opt._numdiff.approx_derivative(fcn,x0,method='2-point',bounds=(lb,ub))
+    try:
+        J = opt._numdiff.approx_derivative(fcn,x0,method='2-point',bounds=(lb,ub))
+    except ValueError as err:
+        if '`x0` violates bound constraints.' in str(err):
+            # Protect against seldom float-point round-off errors
+            x0,lb,ub = [var.round(12) for var in [x0,lb,ub]]
```
Comment (stestoll) [inline deerlab/utils/utils.py:416]: This looks like a hack. What are the conditions that trigger this error?

File: deerlab/utils/utils.py [hunk 2]
```diff
@@ -408,7 +408,15 @@ def Jacobian(fcn, x0, lb, ub):
     This is a wrapper around the ``scipy.optimize._numdiff.approx_derivative`` function.
 
     """
-    J = opt._numdiff.approx_derivative(fcn,x0,method='2-point',bounds=(lb,ub))
+    try:
+        J = opt._numdiff.approx_derivative(fcn,x0,method='2-point',bounds=(lb,ub))
+    except ValueError as err:
```
Comment (stestoll) [inline deerlab/utils/utils.py:413]: Add a test where this if branch gets triggered.

**Review by stestoll — CHANGES_REQUESTED** [review]

File: deerlab/utils/utils.py (hunk 1 above)
Comment (luisfabib) [reply deerlab/utils/utils.py:416]: This is casused by a round-off error. During the Jacobian evulation for the uncertianty propagation one of the values is evaluated at `x=0.9999999999999964` but its bounds are `[1.0,inf]` leading to `x<lb` and to the error by `scipy.optimize._numdiff.approx_derivative`

_Originally posted by @luisfabib in https://github.com/JeschkeLab/DeerLab/issues/177#issuecomment-868276540_

**Comment by luisfabib** [comment]
After digging a bit more, the float-point error appears in one of the variables returned by the `scipy.optimize.lsq_linear` function. I removed the changed to `Jacobian` and added an explicit control to the results of the linear LSQ results in `snlls`.

**Review by stestoll — APPROVED** [review]

**Pull request merged**